File: c/mt76_histogram.c
/* SPDX-License-Identifier: BSD-3-Clause-Clear */
#include "mt76_histogram.h"
#include <string.h>

static uint32_t le32(const uint8_t *p) {
    return (uint32_t)p[0]|(uint32_t)p[1]<<8|(uint32_t)p[2]<<16|(uint32_t)p[3]<<24;
}
static int body(int chip, const uint8_t *raw, size_t len, unsigned eid, unsigned seq) {
    if (chip!=MT_CHIP_MT7925 || !raw || len<44) return -1;
    uint32_t word=le32(raw), size=word&65535;
    if (size<44 || size>len || word>>27!=7 || ((word>>16)&15)==1 || raw[36]!=eid || raw[37]!=seq) return -1;
    return (int)size-44;
}
int mt_histogram_request(int chip, uint8_t *out, size_t capacity) {
    const uint8_t request[8]={0,0,0,0,2,0,4,0};
    if (chip!=MT_CHIP_MT7925 || !out || capacity<8) return -1;
    memcpy(out,request,8); return 8;
}
int mt_histogram_ack(int chip, const uint8_t *raw, size_t len, uint8_t sequence, uint32_t *status) {
    if (!status || !sequence || sequence>15 || body(chip,raw,len,1,sequence)!=8 || le32(raw+44)!=0x36) return -1;
    *status=le32(raw+48); return 0;
}
static void fill(mt_histogram_bins_t *out, const uint8_t *raw) {
    const int8_t labels[10]={-92,-89,-86,-83,-80,-75,-70,-65,-60,-55};
    memcpy(out->threshold_labels_raw,labels,10);
    for (unsigned view=0;view<out->view_count;view++)
        for (unsigned i=0;i<11;i++) {
            out->bins[view][i]=le32(raw+44*view+4*i);
            out->totals[view]+=out->bins[view][i];
        }
}
int mt_histogram_event(int chip, const uint8_t *raw, size_t len, mt_histogram_bins_t *out) {
    const uint8_t prefix[8]={0,0,0,0,2,0,92,0};
    if (!out || body(chip,raw,len,0x36,0)!=96 || memcmp(raw+44,prefix,8)) return -1;
    mt_histogram_bins_t result={.chip=chip,.source=MT_HISTOGRAM_FIRMWARE_TIMER,.view_count=2};
    fill(&result,raw+52); *out=result; return 0;
}
int mt_histogram_legacy(int chip, const uint8_t *raw, size_t len, mt_histogram_bins_t *out) {
    if (chip!=MT_CHIP_MT7921 || !raw || !out || len!=44) return -1;
    mt_histogram_bins_t result={.chip=chip,.source=MT_HISTOGRAM_LEGACY_ORDINARY,.view_count=1};
    fill(&result,raw); *out=result; return 0;
}

File: mt76_histogram.py
# SPDX-License-Identifier: BSD-3-Clause-Clear
"""Experimental raw histogram wire records, not calibrated power or occupancy.

Acquisition resets shared history and needs a separately owned/restored lifetime.
These pure helpers neither start hardware nor establish freshness or coverage.
MT7921's ordinary bank and MT7925's two timer views are different profiles.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

THRESHOLD_LABELS_RAW = (-92, -89, -86, -83, -80, -75, -70, -65, -60, -55)


@dataclass(frozen=True)
class HistogramBins:
    chip: str
    source: str
    bins: tuple[tuple[int, ...], ...]
    threshold_labels_raw: tuple[int, ...] = THRESHOLD_LABELS_RAW

    @property
    def totals(self) -> tuple[int, ...]:
        """Sums of collected samples, without32-bit accumulator overflow."""
        return tuple(sum(view) for view in self.bins)


def build_histogram_request(chip: str) -> bytes:
    """MT7925 UNI36/tag2 one-shot; no duration/mask/index override."""
    if chip != "mt7925":
        raise ValueError("histogram event request supports MT7925 only")
    return struct.pack("<4xHH", 2, 4)


def _body(chip, raw, eid, sequence):
    if chip != "mt7925" or len(raw) < 44:
        raise ValueError("histogram event supports complete MT7925 records only")
    word = struct.unpack_from("<I", raw)[0]
    size = word & 65535
    if (
        not 44 <= size <= len(raw)
        or word >> 27 != 7
        or (word >> 16) & 15 == 1
        or raw[36] != eid
        or raw[37] != sequence
    ):
        raise ValueError("histogram DMA/type/sequence mismatch")
    return raw[44:size]


def parse_histogram_ack(chip: str, raw: bytes, sequence: int) -> int:
    if type(sequence) is not int or not 1 <= sequence <= 15:
        raise ValueError("histogram ACK requires nonzero sequence1..15")
    body = _body(chip, raw, 1, sequence)
    if len(body) != 8 or struct.unpack_from("<I", body)[0] != 0x36:
        raise ValueError("histogram ACK shape/CID mismatch")
    return struct.unpack_from("<I", body, 4)[0]


def parse_histogram_event(chip: str, raw: bytes) -> HistogramBins:
    """Strict EID36/seq0 tag2/len92: two raw views, not antenna labels.

    All-zero and unequal-total arrays are retained, not repaired or labeled
    unavailable. Valid syntax does not establish exposure time or sensor health.
    """
    body = _body(chip, raw, 0x36, 0)
    if len(body) != 96 or body[:8] != struct.pack("<4xHH", 2, 92):
        raise ValueError("histogram event shape mismatch")
    return HistogramBins(
        chip,
        "firmware_timer",
        (struct.unpack_from("<11I", body, 8), struct.unpack_from("<11I", body, 52)),
    )


def parse_legacy_histogram(chip: str, raw: bytes) -> HistogramBins:
    """Exactly eleven little-endian words from the stopped MT7921 ordinary bank.

    No USB read, freeze, reset or hardware ownership is performed here.
    """
    if chip != "mt7921" or len(raw) != 44:
        raise ValueError("legacy histogram requires MT7921 and exactly44 bank bytes")
    return HistogramBins(chip, "legacy_ordinary", (struct.unpack("<11I", raw),))
